### PR TITLE
convert constants (mostly filenames and extensions) in global to strings

### PR DESCRIPTION
**Pull request by benjones**
Reduce .toDString noise lots of places.

### CODE REVIEW

File: src/dmd/glue.d [hunk 1]
```diff
@@ -164,7 +164,7 @@ void obj_write_deferred(Library library)
         uint hash = 0;
         for (const(char)* p = s.toChars(); *p; p++)
             hash += *p;
-        namebuf.printf("%s_%x_%x.%s", fname, count, hash, global.obj_ext);
+        namebuf.printf("%s_%x_%x.%s", fname, count, hash, global.obj_ext.ptr);
```
Comment (jacob-carlborg) [inline src/dmd/glue.d]: Please pass the length and the pointer separately.

File: src/dmd/link.d [hunk 2]
```diff
@@ -254,27 +254,30 @@ public int runLINK()
...
             if (!linkcmd)
                 linkcmd = vsopt.linkerPath(global.params.is64bit);
 
-            const int status = executecmd(linkcmd, p);
+            const int status = executecmd(linkcmd, p.ptr);
             if (lnkfilename)
             {
-                remove(lnkfilename);
-                FileName.free(lnkfilename);
+                lnkfilename.toCStringThen!((lf) => remove(lf.ptr));
```
Comment (jacob-carlborg) [inline src/dmd/link.d]: Parentheses around `lf` are not necessary.

File: src/dmd/mars.d [hunk 3]
```diff
@@ -756,7 +756,7 @@ private int tryMain(size_t argc, const(char)** argv, ref Param params)
                     if (params.oneobj)
                         break;
                 }
-                remove(params.exefile);
+                params.exefile.toCStringThen!((ef) => File.remove(ef.ptr));
```
Comment (jacob-carlborg) [inline src/dmd/mars.d]: Same here, parentheses around `ef` are not necessary.

File: src/dmd/mars.d [hunk 4]
```diff
@@ -2674,20 +2674,20 @@ Modules createModules(ref Strings files, ref Strings libmodules)
             }
             else
             {
-                error(Loc.initial, "unrecognized file extension %s", ext);
+                error(Loc.initial, "unrecognized file extension %s", ext.ptr);
```
Comment (jacob-carlborg) [inline src/dmd/mars.d]: Please pass the length and the pointer separately.

**Comment by jacob-carlborg** [comment]
Looks good. Just a few minor things.

**Comment by benjones** [comment]
Any idea what the shell32.lib failure on windows is about?

**Comment by jacob-carlborg** [comment]
> Any idea what the shell32.lib failure on windows is about?

Sorry, no.

**Comment by benjones** [comment]
Looks like it was an issue with.... null termination.  Looks like everything's running now.

File: src/dmd/mars.d [hunk 5]
```diff
@@ -2674,20 +2674,20 @@ Modules createModules(ref Strings files, ref Strings libmodules)
             }
             else
             {
-                error(Loc.initial, "unrecognized file extension %s", ext);
+                error(Loc.initial, "unrecognized file extension %.*s", ext.length, ext.ptr);
```
Comment (jacob-carlborg) [inline src/dmd/mars.d]: The length is supposed to be of type `int`.

File: src/dmd/glue.d [hunk 6]
```diff
@@ -164,7 +164,7 @@ void obj_write_deferred(Library library)
         uint hash = 0;
         for (const(char)* p = s.toChars(); *p; p++)
             hash += *p;
-        namebuf.printf("%s_%x_%x.%s", fname, count, hash, global.obj_ext);
+        namebuf.printf("%s_%x_%x.%.*s", fname, count, hash, global.obj_ext.length, global.obj_ext.ptr);
```
Comment (jacob-carlborg) [inline src/dmd/glue.d]: The length is supposed to be of type `int`.

**Review by jacob-carlborg — DISMISSED** [review]

File: src/dmd/mars.d (hunk 5 above)
Comment (benjones) [reply src/dmd/mars.d]: So just add cast(int) ?

Comment (jacob-carlborg) [reply src/dmd/mars.d]: Yes.

**Review by thewilsonator — APPROVED** [review]

**Pull request merged**